mash/mash.go

```diff
@@ -0,0 +1,92 @@
+/*
+Package mash is for sketching sequence data to make it easier to search for and against.
+
+The package is named mash after the mash sketching algorithm, which is based on the MinHash algorithm.
+
+Mash: fast genome and metagenome distance estimation using MinHash.
+Ondov, B.D., Treangen, T.J., Melsted, P. et al.
+Genome Biol 17, 132 (2016).
+https://doi.org/10.1186/s13059-016-0997-x
+
+Mash Screen: high-throughput sequence containment estimation for genome discovery.
+Ondov, B., Starrett, G., Sappington, A. et al.
+Genome Biol 20, 232 (2019).
+https://doi.org/10.1186/s13059-019-1841-x
+
+The idea is that we can sketch a sequence of data, and then compare the sketch to other sketches to see how similar they are.
+This saves a bunch of computation time and memory, because we don't have to compare the entire sequence to another sequence.
+
+TTFN,
+Tim
+*/
+package mash
+
+import "github.com/spaolacci/murmur3"
+
+// sketch algorithm
+// slide a window of size k along the sequence
+// for each kmer in the window, hash it to a 32 or 64 bit number
+// keep the minimum hash value of all the kmers in the window up to a given sketch size
+// the sketch is a vector of the minimum hash values
+
+// what are mash's inputs and outputs?
+// inputs: a
+
+type Mash struct {
+	KmerSize   uint
+	SketchSize uint
+	Sketches   []uint32
+}
+
+func NewMash(kmerSize uint, sketchSize uint) *Mash {
+	return &Mash{
+		KmerSize:   kmerSize,
+		SketchSize: sketchSize,
+		Sketches:   make([]uint32, sketchSize),
+	}
+}
+
+func (m *Mash) Sketch(sequence string) {
+	// slide a window of size k along the sequence
+	for kmerStart := 0; kmerStart < len(sequence)-int(m.KmerSize); kmerStart++ {
+		kmer := sequence[kmerStart : kmerStart+int(m.KmerSize)]
+		// hash the kmer to a 32 bit number
+		hash := murmur3.Sum32([]byte(kmer))
+		// keep the minimum hash value of all the kmers in the window up to a given sketch size
+		// the sketch is a vector of the minimum hash values
+		var biggestHashIndex int
+
+		// find the biggest hash value in the sketch
```

Comment:
OH THIS is how it works. Basically, you hash each kmer window, and then keep a list of hashes, small to large. The shared hashes are then compared. There should be a comment somewhere generally explaining the concept behind this in a bit more detail than just the sketches.

Reply:
Yeah that's more or less how it works.

Keep it mind that sketches generated via different kmer window sizes are non-comparable. That's why I thought to store kmer size in the Mash struct @soypat because it could be relevant later. Open to better ways/ more idiomatic ways of doing this.

Reply:
If Kmer size is important then changes I proposed are probably not the best way to go. Knowing this I'm thinking the following usage would be nice:

```go
m1, err :=mash.NewMash(seq1, MashConfig{Sketches: 23, KmerSize:k1})
// handle your errors...
m2, err := mash.NewMash(seq2,  MashConfig{Sketches: 23, KmerSize:k2})

 // Don't let users modify kmerSize field since it "invalidates" calculated data, do expose it via a method.
if m1.KmerSize() != m2.KmerSize() {
    panic("oh no!")
}

if mash.Equal(m1, m2) {
   print("mashes equal!")
}
```

Reply:
@soypat think I'm going to try writing it like this today. Mashes are measured by their Jaccard Similarity (total intersection of hashes / actual hashes so instead will likely make it `mash.Simularity` and have it return a floating point between 0-1.

One other thing I can probably handle by myself is keeping the hashes organized by lexical least to lexically greatest during creation. Any advice on the fastest/most idiomatic way of doing this in Go?